packages/logging/src/analytics-helpers.ts

```diff
@@ -108,15 +148,153 @@ export class ToggleableAnalytics implements MongoshAnalytics {
     if (!('userId' in firstArg && firstArg.userId) &&
         !('anonymousId' in firstArg && firstArg.anonymousId)) {
       const err = new Error('Telemetry setup is missing userId or anonymousId');
-      switch (this._state) {
-        case 'enabled':
-          throw err;
-        case 'paused':
-          this._pendingError ??= err;
-          break;
-        default:
-          break;
+      if (this._queue.isPaused()) {
+        this._pendingError ??= err;
+      } else {
+        throw err;
+      }
+    }
+  }
+
+  flush(callback: (err?: Error | undefined) => void): void {
+    return this._target.flush(callback);
+  }
+}
+
+type ThrottledAnalyticsOptions = {
+  target: MongoshAnalytics;
+  /**
+   * Throttling options. If not provided, throttling is disabled (default: null)
+   */
+  throttle: {
+    /** Allowed events per timeframe number */
+    rate: number;
+    /** Timeframe for throttling (default: 60_000ms) */
+    timeframe?: number;
+    /** Path to persist rpm value to be able to track them between sessions */
+    metadataPath: string;
+  } | null;
+};
+
+export class ThrottledAnalytics implements MongoshAnalytics {
+  private trackQueue = new Queue<AnalyticsTrackMessage>((message) => {
+    if (this.shouldEmitAnalyticsEvent()) {
+      this.target.track(message);
+      this.throttleState.count++;
+    }
+  });
+  private target: ThrottledAnalyticsOptions['target'] = new NoopAnalytics();
+  private currentUserId: string | null = null;
```

Comment:
Is `currentUserId` just being used to track whether telemetry had been enabled before? We don’t actually care about the event that different mongosh instances with different user ids use the same local data path, right?

Reply:
No, no strong preferences, just wanted to make sure I got this right :) I don’t think it matters much in practice as you said, and even if we had multiple users writing to the same data path, I think “throttle per machine” rather than “throttle per user” would not really be worse. But nothing to do about this here.

(This code here _does_ make the assumption that the problem doesn’t occur because of concurrent mongosh sessions, right? Do we want to account for those in some way?)

Reply:
Ohhh, that's a very good point that I totally missed, yeah, we should account for that

Reply:
As discussed, we decided that for concurrent case we will allow only one process to send telemetry. Added lock implementation in https://github.com/mongodb-js/mongosh/pull/1475/commits/c8b28f661b61f5433c83024f6f2915bc262dc03a